File: js/src/experimental/sandbox/sandbox.ts
/**
 * Sandbox class for interacting with a specific sandbox instance.
 */

import type { SandboxClient } from "./client.js";
import type { ExecutionResult, RunOptions, SandboxData } from "./types.js";
import { LangSmithDataplaneNotConfiguredError } from "./errors.js";
import { handleSandboxHttpError } from "./helpers.js";

/**
 * Represents an active sandbox for running commands and file operations.
 *
 * This class is typically obtained from SandboxClient.createSandbox() and
 * provides methods for command execution and file I/O within the sandbox
 * environment.
 *
 * @example
 * ```typescript
 * const sandbox = await client.createSandbox("python-sandbox");
 * try {
 *   const result = await sandbox.run("python --version");
 *   console.log(result.stdout);
 * } finally {
 *   await sandbox.delete();
 * }
 * ```
 */
export class Sandbox {
  /** Display name (can be updated). */
  readonly name: string;
  /** Name of the template used to create this sandbox. */
  readonly template_name: string;
  /** URL for data plane operations (file I/O, command execution). */
  readonly dataplane_url?: string;
  /** Unique identifier (UUID). Remains constant even if name changes. */
  readonly id?: string;
  /** Timestamp when the sandbox was created. */
  readonly created_at?: string;
  /** Timestamp when the sandbox was last updated. */
  readonly updated_at?: string;

  private _client: SandboxClient;

  /** @internal */
  constructor(data: SandboxData, client: SandboxClient) {
    this.name = data.name;
    this.template_name = data.template_name;
    this.dataplane_url = data.dataplane_url;
    this.id = data.id;
    this.created_at = data.created_at;
    this.updated_at = data.updated_at;
    this._client = client;
  }

  /**
   * Validate and return the dataplane URL.
   * @throws LangSmithDataplaneNotConfiguredError if dataplane_url is not configured.
   */
  private requireDataplaneUrl(): string {
    if (!this.dataplane_url) {
      throw new LangSmithDataplaneNotConfiguredError(
        `Sandbox '${this.name}' does not have a dataplane_url configured. ` +
          "Runtime operations require a dataplane URL."
      );
    }
    return this.dataplane_url;
  }

  /**
   * Execute a command in the sandbox.
   *
   * @param command - Shell command to execute.
   * @param options - Execution options.
   * @returns ExecutionResult with stdout, stderr, and exit_code.
   * @throws LangSmithDataplaneNotConfiguredError if dataplane_url is not configured.
   * @throws SandboxOperationError if command execution fails.
   * @throws SandboxConnectionError if connection to sandbox fails.
   * @throws SandboxNotReadyError if sandbox is not ready.
   *
   * @example
   * ```typescript
   * const result = await sandbox.run("echo hello");
   * console.log(result.stdout); // "hello\n"
   * console.log(result.exit_code); // 0
   * ```
   */
  async run(
    command: string,
    options: RunOptions = {}
  ): Promise<ExecutionResult> {
    const { timeout = 60, env, cwd, shell = "/bin/bash" } = options;
    const dataplaneUrl = this.requireDataplaneUrl();
    const url = `${dataplaneUrl}/execute`;

    const payload: Record<string, unknown> = {
      command,
      timeout,
      shell,
    };
    if (env !== undefined) {
      payload.env = env;
    }
    if (cwd !== undefined) {
      payload.cwd = cwd;
    }

    const response = await this._client._fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout((timeout + 10) * 1000),
    });

    if (!response.ok) {
      await handleSandboxHttpError(response);
    }

    const data = await response.json();
    return {
      stdout: data.stdout ?? "",
      stderr: data.stderr ?? "",
      exit_code: data.exit_code ?? -1,
    };
  }

  /**
   * Write content to a file in the sandbox.
   *
   * @param path - Target file path in the sandbox.
   * @param content - File content (string or bytes).
   * @param timeout - Request timeout in seconds.
   * @throws LangSmithDataplaneNotConfiguredError if dataplane_url is not configured.
   * @throws SandboxOperationError if file write fails.
   * @throws SandboxConnectionError if connection to sandbox fails.
   *
   * @example
   * ```typescript
   * await sandbox.write("/tmp/script.py", 'print("Hello!")');
   * ```
   */
  async write(
    path: string,
    content: string | Uint8Array,
    timeout = 60
  ): Promise<void> {
    const dataplaneUrl = this.requireDataplaneUrl();
    const url = `${dataplaneUrl}/upload?path=${encodeURIComponent(path)}`;

    // Ensure content is bytes for multipart upload
    const bytes =
      typeof content === "string" ? new TextEncoder().encode(content) : content;

    const formData = new FormData();
    // Create a copy to ensure we have a plain ArrayBuffer (not SharedArrayBuffer)
    const buffer = new Uint8Array(bytes).buffer as ArrayBuffer;
    const blob = new Blob([buffer], { type: "application/octet-stream" });
    formData.append("file", blob, "file");

    const response = await this._client._fetch(url, {
      method: "POST",
      body: formData,
      signal: AbortSignal.timeout(timeout * 1000),
    });

    if (!response.ok) {
      await handleSandboxHttpError(response);
    }
  }

  /**
   * Read a file from the sandbox.
   *
   * @param path - File path to read. Supports both absolute paths (e.g., /tmp/file.txt)
   *               and relative paths (resolved from /home/user/).
   * @param timeout - Request timeout in seconds.
   * @returns File contents as Uint8Array.
   * @throws LangSmithDataplaneNotConfiguredError if dataplane_url is not configured.
   * @throws ResourceNotFoundError if the file doesn't exist.
   * @throws SandboxOperationError if file read fails.
   * @throws SandboxConnectionError if connection to sandbox fails.
   *
   * @example
   * ```typescript
   * const content = await sandbox.read("/tmp/output.txt");
   * const text = new TextDecoder().decode(content);
   * console.log(text);
   * ```
   */
  async read(path: string, timeout = 60): Promise<Uint8Array> {
    const dataplaneUrl = this.requireDataplaneUrl();
    const url = `${dataplaneUrl}/download?path=${encodeURIComponent(path)}`;

    const response = await this._client._fetch(url, {
      method: "GET",
      signal: AbortSignal.timeout(timeout * 1000),
    });

    if (!response.ok) {
      await handleSandboxHttpError(response);
    }

    const buffer = await response.arrayBuffer();
    return new Uint8Array(buffer);
  }

  /**
   * Delete this sandbox.
   *
   * Call this when you're done using the sandbox to clean up resources.
   *
   * @example
   * ```typescript
   * const sandbox = await client.createSandbox("python-sandbox");
   * try {
   *   await sandbox.run("echo hello");
   * } finally {
   *   await sandbox.delete();
   * }
   * ```
   */
  async delete(): Promise<void> {
    await this._client.deleteSandbox(this.name);
  }
}

File: python/langsmith/sandbox/_sandbox.py
"""Sandbox class for interacting with a specific sandbox instance."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable, Literal, Optional, Union, overload

import httpx

from langsmith.sandbox._exceptions import (
    DataplaneNotConfiguredError,
    ResourceNotFoundError,
    SandboxConnectionError,
)
from langsmith.sandbox._helpers import handle_sandbox_http_error
from langsmith.sandbox._models import (
    CommandHandle,
    ExecutionResult,
)

if TYPE_CHECKING:
    from langsmith.sandbox._client import SandboxClient


@dataclass
class Sandbox:
    """Represents an active sandbox for running commands and file operations.

    This class is typically obtained from SandboxClient.sandbox() and supports
    the context manager protocol for automatic cleanup.

    Attributes:
        name: Display name (can be updated).
        template_name: Name of the template used to create this sandbox.
        dataplane_url: URL for data plane operations (file I/O, command execution).
        id: Unique identifier (UUID). Remains constant even if name changes.
            May be None for resources created before ID support was added.
        created_at: Timestamp when the sandbox was created.
        updated_at: Timestamp when the sandbox was last updated.

    Example:
        with client.sandbox(template_name="python-sandbox") as sandbox:
            result = sandbox.run("python --version")
            print(result.stdout)
    """

    # Data fields (from API response)
    name: str
    template_name: str
    dataplane_url: Optional[str] = None
    id: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

    # Internal fields (not from API)
    _client: SandboxClient = field(repr=False, default=None)  # type: ignore
    _auto_delete: bool = field(repr=False, default=True)

    @classmethod
    def from_dict(
        cls,
        data: dict[str, Any],
        client: SandboxClient,
        auto_delete: bool = True,
    ) -> Sandbox:
        """Create a Sandbox from API response dict.

        Args:
            data: API response dictionary containing sandbox data.
            client: Parent SandboxClient for operations.
            auto_delete: Whether to delete the sandbox on context exit.

        Returns:
            Sandbox instance.
        """
        return cls(
            name=data.get("name", ""),
            template_name=data.get("template_name", ""),
            dataplane_url=data.get("dataplane_url"),
            id=data.get("id"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
            _client=client,
            _auto_delete=auto_delete,
        )

    def __enter__(self) -> Sandbox:
        """Enter context manager."""
        return self

    def __exit__(
        self,
        exc_type: Optional[type],
        exc_val: Optional[BaseException],
        exc_tb: Optional[Any],
    ) -> None:
        """Exit context manager, optionally deleting the sandbox."""
        if self._auto_delete:
            try:
                self._client.delete_sandbox(self.name)
            except Exception:
                # Don't raise on cleanup errors
                pass

    def _require_dataplane_url(self) -> str:
        """Validate and return the dataplane URL.

        Returns:
            The dataplane URL.

        Raises:
            DataplaneNotConfiguredError: If dataplane_url is not configured.
        """
        if not self.dataplane_url:
            raise DataplaneNotConfiguredError(
                f"Sandbox '{self.name}' does not have a dataplane_url configured. "
                "Runtime operations require a dataplane URL."
            )
        return self.dataplane_url

    @overload
    def run(
        self,
        command: str,
        *,
        timeout: int = ...,
        env: Optional[dict[str, str]] = ...,
        cwd: Optional[str] = ...,
        shell: str = ...,
        on_stdout: Optional[Callable[[str], Any]] = ...,
        on_stderr: Optional[Callable[[str], Any]] = ...,
        wait: Literal[True] = ...,
    ) -> ExecutionResult: ...

    @overload
    def run(
        self,
        command: str,
        *,
        timeout: int = ...,
        env: Optional[dict[str, str]] = ...,
        cwd: Optional[str] = ...,
        shell: str = ...,
        on_stdout: Optional[Callable[[str], Any]] = ...,
        on_stderr: Optional[Callable[[str], Any]] = ...,
        wait: Literal[False],
    ) -> CommandHandle: ...

    def run(
        self,
        command: str,
        *,
        timeout: int = 60,
        env: Optional[dict[str, str]] = None,
        cwd: Optional[str] = None,
        shell: str = "/bin/bash",
        on_stdout: Optional[Callable[[str], Any]] = None,
        on_stderr: Optional[Callable[[str], Any]] = None,
        wait: bool = True,
    ) -> Union[ExecutionResult, CommandHandle]:
        """Execute a command in the sandbox.

        Args:
            command: Shell command to execute.
            timeout: Command timeout in seconds.
            env: Environment variables to set for the command.
            cwd: Working directory for command execution. If None, uses sandbox default.
            shell: Shell to use for command execution. Defaults to "/bin/bash".
            on_stdout: Callback invoked with each stdout chunk as it arrives.
                Blocks until the command completes and returns ExecutionResult.
                Cannot be combined with wait=False.
            on_stderr: Callback invoked with each stderr chunk as it arrives.
                Blocks until the command completes and returns ExecutionResult.
                Cannot be combined with wait=False.
            wait: If True (default), block until the command completes and
                return ExecutionResult. If False, return a
                CommandHandle immediately for streaming output,
                kill, stdin input, and reconnection. Cannot be combined with
                on_stdout/on_stderr callbacks.

        Returns:
            ExecutionResult when wait=True (default).
            CommandHandle when wait=False.

        Raises:
            ValueError: If wait=False is combined with callbacks.
            DataplaneNotConfiguredError: If dataplane_url is not configured.
            SandboxOperationError: If command execution fails.
            CommandTimeoutError: If command exceeds its timeout.
            SandboxConnectionError: If connection to sandbox fails.
            SandboxNotReadyError: If sandbox is not ready.
            SandboxClientError: For other errors.
        """
        if not wait and (on_stdout or on_stderr):
            raise ValueError(
                "Cannot combine wait=False with on_stdout/on_stderr callbacks. "
                "Use wait=False and iterate the CommandHandle, or use callbacks."
            )

        self._require_dataplane_url()

        # When not waiting or callbacks are requested, WS is required
        use_ws = not wait or on_stdout or on_stderr
        if use_ws:
            return self._run_ws(
                command,
                timeout=timeout,
                env=env,
                cwd=cwd,
                shell=shell,
                wait=wait,
                on_stdout=on_stdout,
                on_stderr=on_stderr,
            )

        # Default (wait=True, no callbacks): try WS, fall back to HTTP.
        # Catch broad exceptions so that unexpected WS failures (e.g. version
        # incompatibilities) don't break users who don't need WS features.
        try:
            return self._run_ws(
                command,
                timeout=timeout,
                env=env,
                cwd=cwd,
                shell=shell,
                wait=True,
                on_stdout=None,
                on_stderr=None,
            )
        except (SandboxConnectionError, ImportError, OSError, TypeError):
            return self._run_http(
                command,
                timeout=timeout,
                env=env,
                cwd=cwd,
                shell=shell,
            )

    def _run_ws(
        self,
        command: str,
        *,
        timeout: int,
        env: Optional[dict[str, str]],
        cwd: Optional[str],
        shell: str,
        wait: bool,
        on_stdout: Optional[Callable[[str], Any]],
        on_stderr: Optional[Callable[[str], Any]],
    ) -> Union[ExecutionResult, CommandHandle]:
        """Execute via WebSocket /execute/ws."""
        from langsmith.sandbox._ws_execute import run_ws_stream

        dataplane_url = self._require_dataplane_url()
        api_key = self._client._api_key

        msg_stream, control = run_ws_stream(
            dataplane_url,
            api_key,
            command,
            timeout=timeout,
            env=env,
            cwd=cwd,
            shell=shell,
            on_stdout=on_stdout,
            on_stderr=on_stderr,
        )

        handle = CommandHandle(msg_stream, control, self)

        if not wait:
            return handle

        return handle.result  # blocks until command completes

    def _run_http(
        self,
        command: str,
        *,
        timeout: int,
        env: Optional[dict[str, str]],
        cwd: Optional[str],
        shell: str,
    ) -> ExecutionResult:
        """Execute via HTTP POST /execute (existing implementation)."""
        dataplane_url = self._require_dataplane_url()
        url = f"{dataplane_url}/execute"
        payload: dict[str, Any] = {
            "command": command,
            "timeout": timeout,
            "shell": shell,
        }
        if env is not None:
            payload["env"] = env
        if cwd is not None:
            payload["cwd"] = cwd

        try:
            response = self._client._http.post(url, json=payload, timeout=timeout + 10)
            response.raise_for_status()
            data = response.json()
            return ExecutionResult(
                stdout=data.get("stdout", ""),
                stderr=data.get("stderr", ""),
                exit_code=data.get("exit_code", -1),
            )
        except httpx.ConnectError as e:
            raise SandboxConnectionError(
                f"Failed to connect to sandbox '{self.name}': {e}"
            ) from e
        except httpx.HTTPStatusError as e:
            handle_sandbox_http_error(e)
            # This line should never be reached but satisfies type checker
            raise  # pragma: no cover

    def reconnect(
        self,
        command_id: str,
        *,
        stdout_offset: int = 0,
        stderr_offset: int = 0,
    ) -> CommandHandle:
        """Reconnect to a running or recently-finished command.

        Resumes output from the given byte offsets. Any output produced while
        the client was disconnected is replayed from the server's ring buffer.

        Args:
            command_id: The command ID from handle.command_id.
            stdout_offset: Byte offset to resume stdout from (default: 0).
            stderr_offset: Byte offset to resume stderr from (default: 0).

        Returns:
            A CommandHandle for the command.

        Raises:
            SandboxOperationError: If command_id is not found or session expired.
            SandboxConnectionError: If connection to sandbox fails.
        """
        from langsmith.sandbox._ws_execute import reconnect_ws_stream

        dataplane_url = self._require_dataplane_url()
        api_key = self._client._api_key

        msg_stream, control = reconnect_ws_stream(
            dataplane_url,
            api_key,
            command_id,
            stdout_offset=stdout_offset,
            stderr_offset=stderr_offset,
        )

        return CommandHandle(
            msg_stream,
            control,
            self,
            command_id=command_id,
            stdout_offset=stdout_offset,
            stderr_offset=stderr_offset,
        )

    def write(
        self,
        path: str,
        content: Union[str, bytes],
        *,
        timeout: int = 60,
    ) -> None:
        """Write content to a file in the sandbox.

        Args:
            path: Target file path in the sandbox.
            content: File content (str or bytes).
            timeout: Request timeout in seconds.

        Raises:
            DataplaneNotConfiguredError: If dataplane_url is not configured.
            SandboxOperationError: If file write fails.
            SandboxConnectionError: If connection to sandbox fails.
            SandboxNotReadyError: If sandbox is not ready.
            SandboxClientError: For other errors.
        """
        dataplane_url = self._require_dataplane_url()
        url = f"{dataplane_url}/upload"

        # Ensure content is bytes for multipart upload
        if isinstance(content, str):
            content = content.encode("utf-8")

        files = {"file": ("file", content)}

        try:
            response = self._client._http.post(
                url, params={"path": path}, files=files, timeout=timeout
            )
            response.raise_for_status()
        except httpx.ConnectError as e:
            raise SandboxConnectionError(
                f"Failed to connect to sandbox '{self.name}': {e}"
            ) from e
        except httpx.HTTPStatusError as e:
            handle_sandbox_http_error(e)

    def read(self, path: str, *, timeout: int = 60) -> bytes:
        """Read a file from the sandbox.

        Args:
            path: File path to read. Supports both absolute paths (e.g., /tmp/file.txt)
                  and relative paths (resolved from /home/user/).
            timeout: Request timeout in seconds.

        Returns:
            File contents as bytes.

        Raises:
            DataplaneNotConfiguredError: If dataplane_url is not configured.
            ResourceNotFoundError: If the file doesn't exist.
            SandboxOperationError: If file read fails.
            SandboxConnectionError: If connection to sandbox fails.
            SandboxNotReadyError: If sandbox is not ready.
            SandboxClientError: For other errors.
        """
        dataplane_url = self._require_dataplane_url()
        url = f"{dataplane_url}/download"

        try:
            response = self._client._http.get(
                url, params={"path": path}, timeout=timeout
            )
            response.raise_for_status()
            return response.content
        except httpx.ConnectError as e:
            raise SandboxConnectionError(
                f"Failed to connect to sandbox '{self.name}': {e}"
            ) from e
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 404:
                raise ResourceNotFoundError(
                    f"File '{path}' not found in sandbox '{self.name}'",
                    resource_type="file",
                ) from e
            handle_sandbox_http_error(e)
            # This line should never be reached but satisfies type checker
            raise  # pragma: no cover
